test/api/test_declaration.py

```diff
@@ -1026,3 +1026,106 @@ async def test_publication_modalités_or_url_is_required_for_2020(client, body):
+    await declaration(
+        siren="514027945",
+        year=2020,
+        owner="foo@bar.org",
+        entreprise={
+            "adresse": "1 rue de Trois",
+            "code_postal": "77480",
+            "commune": "Quatre",
+        },
+    )
+    monkeypatch.setattr("egapro.emails.send", mock_send)
+    resp = await client.post("/declaration/514027945/2020/receipt")
+    assert resp.status == 204
+    assert calls == 1
+
+
+async def test_resend_receipt_endpoint_by_staff(client, monkeypatch, declaration):
+    calls = 0
+
+    def mock_send(to, subject, txt, html, reply_to, attachment):
+        assert to == ["foo@bar.org", "foo@foo.foo"]
+        assert "/declaration/?siren=514027945&year=2020" in txt
+        assert "/declaration/?siren=514027945&year=2020" in html
+        assert attachment[1] == "declaration_514027945_2021.pdf"
+        nonlocal calls
+        calls += 1
+
+    await db.ownership.put("514027945", "foo@bar.org")
+    # Add another owner, that should be in the email recipients
+    await db.ownership.put("514027945", "foo@foo.foo")
+    await declaration(
+        siren="514027945",
+        year=2020,
+        owner="foo@bar.org",
+        entreprise={
+            "adresse": "1 rue de Trois",
+            "code_postal": "77480",
+            "commune": "Quatre",
+        },
+    )
+    monkeypatch.setattr("egapro.emails.send", mock_send)
+    monkeypatch.setattr("egapro.config.STAFF", ["staff@email.com"])
+    client.login("Staff@email.com")
+    resp = await client.post("/declaration/514027945/2020/receipt")
+    assert resp.status == 204
+    assert calls == 1
+
+
+async def test_resend_receipt_endpoint_by_non_owner(client, monkeypatch, declaration):
+    calls = 0
+
+    def mock_send(to, subject, txt, html, reply_to, attachment):
+        nonlocal calls
+        calls += 1
+
+    await db.ownership.put("514027945", "foo@bar.org")
+    # Add another owner, that should be in the email recipients
+    await db.ownership.put("514027945", "foo@foo.foo")
+    await declaration(
+        siren="514027945",
+        year=2020,
+        owner="foo@bar.org",
+        entreprise={
+            "adresse": "1 rue de Trois",
+            "code_postal": "77480",
+            "commune": "Quatre",
+        },
+    )
+    monkeypatch.setattr("egapro.emails.send", mock_send)
+    client.login("non@owner.com")
+    resp = await client.post("/declaration/514027945/2020/receipt")
+    assert resp.status == 403
+    assert not calls
+
+
+async def test_resend_receipt_endpoint_with_unknown_declaration(client, monkeypatch):
+    calls = 0
+
+    def mock_send(to, subject, txt, html, reply_to, attachment):
+        nonlocal calls
+        calls += 1
+
+    await db.ownership.put("514027945", "foo@bar.org")
+    monkeypatch.setattr("egapro.emails.send", mock_send)
+    resp = await client.post("/declaration/514027945/2019/receipt")
+    assert resp.status == 404
+    assert not calls
```

Comment:
quite a few repetitions in those tests, how about using a Mock object (which you can assert on the number of calls and arguments) like in https://stackoverflow.com/a/67074212, and sharing it with all the tests, would that make sense?

Reply:
:+1: 

Funny thing is that I was doing that in some other places of the file :p